v2/bundle/spiffebundle/bundle.go

```diff
@@ -3,175 +3,411 @@ package spiffebundle
-	panic("not implemented")
+func Parse(trustDomain spiffeid.TrustDomain, bundleBytes []byte) (*Bundle, error) {
+	jwks := &bundleDoc{}
+	if err := json.Unmarshal(bundleBytes, jwks); err != nil {
+		return nil, spiffebundleErr.New("unable to parse JWKS: %v", err)
+	}
+
+	bundle := New(trustDomain)
+	if jwks.RefreshHint > 0 {
+		bundle.refreshHint = &jwks.RefreshHint
+	}
+	if jwks.SequenceNumber > 0 {
+		bundle.sequenceNumber = &jwks.SequenceNumber
+	}
+
+	if jwks.Keys == nil {
+		// The parameter keys MUST be present.
+		// https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE_Trust_Domain_and_Bundle.md#413-keys
+		return nil, spiffebundleErr.New("no keys found")
+	}
+	for i, key := range jwks.Keys {
+		switch key.Use {
+		// Two SVID types are supported: x509-svid and jwt-svid.
+		case x509SVIDUse:
+			if len(key.Certificates) != 1 {
+				return nil, spiffebundleErr.New("expected a single certificate in %s entry %d; got %d", x509SVIDUse, i, len(key.Certificates))
+			}
+			bundle.AddX509Root(key.Certificates[0])
+		case jwtSVIDUse:
+			if err := bundle.AddJWTKey(key.KeyID, key.Key); err != nil {
+				return nil, spiffebundleErr.New("error adding key %d of JWKS: %v", i, errs.Unwrap(err))
+			}
+		case "":
+			// The use parameter MUST be set.
+			// https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE_Trust_Domain_and_Bundle.md#422-public-key-use
+			return nil, spiffebundleErr.New("missing use for key entry %d", i)
+		}
+	}
+
+	return bundle, nil
 }
 
 // FromX509Bundle creates a bundle from an X.509 bundle.
 func FromX509Bundle(x509Bundle *x509bundle.Bundle) *Bundle {
-	panic("not implemented")
+	if x509Bundle != nil {
+		return &Bundle{
+			trustDomain: x509Bundle.TrustDomain(),
+			x509Roots:   x509Bundle.X509Roots(),
+		}
+	}
+	return &Bundle{}
 }
 
 // FromJWTBundle creates a bundle from a JWT bundle.
 func FromJWTBundle(jwtBundle *jwtbundle.Bundle) *Bundle {
-	panic("not implemented")
+	if jwtBundle != nil {
+		return &Bundle{
+			trustDomain: jwtBundle.TrustDomain(),
+			jwtKeys:     jwtBundle.JWTKeys(),
+		}
+	}
+	return &Bundle{}
+}
+
+// FromX509Roots creates a bundle from X.509 certificates.
+func FromX509Roots(trustDomain spiffeid.TrustDomain, x509Roots []*x509.Certificate) *Bundle {
+	return &Bundle{
+		trustDomain: trustDomain,
+		x509Roots:   x509Roots,
+	}
+}
+
+// FromJWTKeys creates a new bundle from JWT public keys.
+func FromJWTKeys(trustDomain spiffeid.TrustDomain, jwtKeys map[string]crypto.PublicKey) *Bundle {
+	return &Bundle{
+		trustDomain: trustDomain,
+		jwtKeys:     jwtKeys,
+	}
 }
 
 // TrustDomain returns the trust domain that the bundle belongs to.
 func (b *Bundle) TrustDomain() spiffeid.TrustDomain {
-	panic("not implemented")
+	return b.trustDomain
 }
 
 // X509Roots returns the X.509 roots in the bundle.
 func (b *Bundle) X509Roots() []*x509.Certificate {
-	panic("not implemented")
+	b.mtx.RLock()
+	defer b.mtx.RUnlock()
+
+	return b.x509Roots
```

Comment:
We should probably copy the slice so we don't accidentally introduce false sharing.